### PR TITLE
[API] [Product Review] Configure product review as a API resource with tests

### PR DESCRIPTION
**Pull request by AdamKasp**
| Q               | A
| --------------- | -----
| Branch?         | api
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| License         | MIT

### CODE REVIEW

File: src/Sylius/Behat/Context/Api/Admin/ManagingProductReviewsContext.php [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+    {
+        $this->assertIfReviewHasElementWithValue($productReview, 'comment', $comment);
+    }
+
+    /**
+     * @Then /^average rating of (product "[^"]+") should be (\d+)$/
+     */
+    public function thisProductAverageRatingShouldBe(ProductInterface $product, int $averageRating): void
+    {
+        // todo, this step should be in product context which will be covered in separate RP
```
Comment (AdamKasp) [inline src/Sylius/Behat/Context/Api/Admin/ManagingProductReviewsContext.php]: after consultation with @Zales0123  I left this step without coverage, this implementation should be in product context which will be configured later

File: src/Sylius/Behat/Resources/config/services/contexts/api.xml [hunk 2]
```diff
@@ -33,6 +33,13 @@
             <argument type="service" id="session" />
         </service>
 
+        <service id="sylius.behat.context.api.admin.managing_product_reviews" class="Sylius\Behat\Context\Api\Admin\ManagingProductReviewsContext">
+            <argument>product_reviews</argument>
```
Comment (AdamKasp) [inline src/Sylius/Behat/Resources/config/services/contexts/api.xml]: IMO resource name we should put as an argument in service, it looks better than many magic strings in code.

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/ProductReview.xml [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+                    <attribute name="groups">product_review:update</attribute>
+                </attribute>
+            </itemOperation>
+            <itemOperation name="accept_product_review">
+                <attribute name="method">PATCH</attribute>
+                <attribute name="path">/product_reviews/{id}/apply_transition/accept</attribute>
+                <attribute name="controller">sylius.api.state_machine_transition_applicator:accept</attribute>
+            </itemOperation>
+            <itemOperation name="reject_product_review">
+                <attribute name="method">PATCH</attribute>
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/ProductReview.xml]: I don't know how configure serialization group to have empty request body in state machine transition paths

File: src/Sylius/Behat/Resources/config/services/contexts/api.xml (hunk 2 above)
Comment (lchrusciel) [reply src/Sylius/Behat/Resources/config/services/contexts/api.xml]: If you want to extract this string, I would rather vote on placing it on the client and create resource-specific clients. However, the idea is nice 👍

File: src/Sylius/Bundle/ApiBundle/Applicator/StateMachineTransitionApplicator.php [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\Applicator;
+
+use SM\Factory\FactoryInterface as StateMachineFactoryInterface;
+use Sylius\Component\Core\Model\ProductReview;
+use Sylius\Component\Core\ProductReviewTransitions;
+use Sylius\Component\Review\Model\ReviewInterface;
+
+final class StateMachineTransitionApplicator
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Applicator/StateMachineTransitionApplicator.php]:
```suggestion
final class ProductReviewStateMachineTransitionApplicator
```

Also, if we want to keep this class this way, I would introduce an interface, or split this class into separate applicators.

File: src/Sylius/Bundle/ApiBundle/Applicator/StateMachineTransitionApplicator.php [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+{
+    /** @var StateMachineFactoryInterface $stateMachineFactory */
+    private $stateMachineFactory;
+
+    public function __construct(StateMachineFactoryInterface $stateMachineFactory)
+    {
+        $this->stateMachineFactory = $stateMachineFactory;
+    }
+
+    public function accept(ProductReview $data): ReviewInterface
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Applicator/StateMachineTransitionApplicator.php]: I don't remember the results, can you remind me what will happen if we will use an interface here?
```suggestion
    public function accept(ReviewInterface $data): ReviewInterface
```

File: src/Sylius/Behat/Context/Api/Admin/ManagingProductReviewsContext.php [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+        $this->client->show($this->resource, (string) $productReview->getId());
+        Assert::true(
+            $this->client->hasValue($element, $value),
+            sprintf('Product review %s is not %s', $element, $value)
+        );
+    }
+
+    private function applyTransition(string $id, string $transition): void
+    {
+        $this->stateMachineClient->applyTransition($this->resource, $id, $transition);
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Api/Admin/ManagingProductReviewsContext.php]: I'm not sure if a specific controller for these kinds of actions is required

File: src/Sylius/Bundle/ApiBundle/Applicator/StateMachineTransitionApplicator.php (hunk 5 above)
Comment (AdamKasp) [reply src/Sylius/Bundle/ApiBundle/Applicator/StateMachineTransitionApplicator.php]: the same result as using class, I will change it to interface

File: src/Sylius/Behat/Context/Api/Admin/ManagingProductReviewsContext.php [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+
+    /**
+     * @When I change its title to :title
+     */
+    public function iChangeItsTitleTo(?string $title = null): void
+    {
+        if ($title !== null) {
+            $this->client->addRequestData('title', $title);
+        }
+    }
```
Comment (Tomanhez) [inline src/Sylius/Behat/Context/Api/Admin/ManagingProductReviewsContext.php]:
```suggestion
    /**
     * @When I change its title to :title
     */
    public function iChangeItsTitleTo(string $title): void
    {
        $this->client->addRequestData('title', $title);
    }
```
You don't need set as null because your step doesn't specify as empty

File: src/Sylius/Behat/Context/Api/Admin/ManagingProductReviewsContext.php [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+
+    /**
+     * @When I change its comment to :comment
+     */
+    public function iChangeItsCommentTo(?string $comment = null): void
+    {
+        if ($comment !== null) {
+            $this->client->addRequestData('comment', $comment);
+        }
+    }
```
Comment (Tomanhez) [inline src/Sylius/Behat/Context/Api/Admin/ManagingProductReviewsContext.php]: As above

File: src/Sylius/Behat/Context/Api/Admin/ManagingProductReviewsContext.php [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+        $this->client->buildUpdateRequest('rating', $rating);
+    }
+
+    /**
+     * @Then I should be editing review of product :productName
+     */
+    public function iShouldBeEditingReviewOfProduct(string $productName): void
+    {
+        $name = $this->client->getCollection()['reviewSubject']['name'];
+        Assert::same($name, $productName);
```
Comment (Tomanhez) [inline src/Sylius/Behat/Context/Api/Admin/ManagingProductReviewsContext.php]:
```suggestion
        Assert::same($this->client->getCollection()['reviewSubject']['name'], $productName);
```
Should enough

File: src/Sylius/Behat/Context/Api/Admin/ManagingProductReviewsContext.php [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+    }
+
+    /**
+     * @Then I should see the customer's name :customerName
+     */
+    public function iShouldSeeTheCustomerSName(string $customerName): void
+    {
+        $name = $this->client->getCollection()['author']['fullName'];
+        Assert::same($name, $customerName);
```
Comment (Tomanhez) [inline src/Sylius/Behat/Context/Api/Admin/ManagingProductReviewsContext.php]: As above

File: src/Sylius/Behat/Client/ApiPlatformStateMachineClient.php [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    public function applyTransition(string $resource, string $id, string $transition): void
+    {
+        $this->client->request(
+            'PATCH',
+            sprintf('/new-api/%s/%s/apply_transition/%s', $resource, $id, $transition),
+            [], [],
+            ['CONTENT_TYPE' => 'application/merge-patch+json', 'HTTP_ACCEPT' => 'application/ld+json'],
+            '{}'
```
Comment (lchrusciel) [inline src/Sylius/Behat/Client/ApiPlatformStateMachineClient.php]: Can you leave a comment, why do we have this string here?

File: src/Sylius/Behat/Context/Api/Admin/ManagingProductReviewsContext.php (hunk 1 above)
Comment (lchrusciel) [reply src/Sylius/Behat/Context/Api/Admin/ManagingProductReviewsContext.php]: IIRC this is the format we try to keep:
```suggestion
        // TODO: this step should be in product context which will be covered in separate RP
```

File: src/Sylius/Behat/Client/ApiPlatformStateMachineClientInterface.php [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Behat\Client;
+
+interface ApiPlatformStateMachineClientInterface
+{
+    public function applyTransition(string $resource, string $id, string $transition): void;
```
Comment (lchrusciel) [inline src/Sylius/Behat/Client/ApiPlatformStateMachineClientInterface.php]: TBH I don't see any difference between `update`, `create` and `applyTransition` in terms of sylius resources from the client perspective. I would add it as the next function to our default client

File: src/Sylius/Behat/Resources/config/services/contexts/api.xml (hunk 2 above)
Comment (lchrusciel) [reply src/Sylius/Behat/Resources/config/services/contexts/api.xml]: If we decide to move into this direction we need an issue for that (or even better a pr, which will update all of currently existing resources)

Comment (AdamKasp) [reply src/Sylius/Behat/Resources/config/services/contexts/api.xml]: I will do this in separate PR

File: src/Sylius/Behat/Client/ApiPlatformClient.php [hunk 13]
```diff
@@ -94,6 +94,17 @@ public function delete(string $resource, string $id): void
...
     }
 
+    public function applyTransition(string $resource, string $id, string $transition): void
+    {
+        $this->client->request(
+            'PATCH',
+            sprintf('/new-api/%s/%s/apply_transition/%s', $resource, $id, $transition),
+            [], [],
+            ['CONTENT_TYPE' => 'application/merge-patch+json', 'HTTP_ACCEPT' => 'application/ld+json'],
+            '{}'
```
Comment (AdamKasp) [inline src/Sylius/Behat/Client/ApiPlatformClient.php]: In this PATCH action, we don't provide any request's content but we call controller.
But we need to provide empty array in this argument because without it we get a syntax error

File: src/Sylius/Behat/Context/Api/Admin/ManagingProductReviewsContext.php [hunk 14]
```diff
@@ -0,0 +1,237 @@
...
+    {
+        $this->client->buildUpdateRequest('product_reviews', (string) $productReview->getId());
+    }
+
+    /**
+     * @When I change its title to :title
+     */
+    public function iChangeItsTitleTo(string $title): void
+    {
+        $this->client->addRequestData('title', $title);
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Api/Admin/ManagingProductReviewsContext.php]: Shouldn't it be `buildUpdateRequest`? Here and below

File: src/Sylius/Behat/Context/Api/Admin/ManagingProductReviewsContext.php [hunk 15]
```diff
@@ -0,0 +1,237 @@
...
+     */
+    public function iChangeItsCommentTo(string $comment): void
+    {
+        $this->client->addRequestData('comment', $comment);
+    }
+
+    /**
+     * @When /^I remove its (comment|title)$/
+     */
+    public function iRemoveItsComment(string $element): void
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Api/Admin/ManagingProductReviewsContext.php]: I would go the other way - make nullable fields in the previous steps

```php
/**
 * @When I change its title to :title
 * @When I remove its title
 */
public function iChangeItsTitleTo(?string $title = ''): void
{
    $this->client->addRequestData('title', $title);
}
```

File: src/Sylius/Behat/Context/Api/Admin/ManagingProductReviewsContext.php [hunk 16]
```diff
@@ -0,0 +1,237 @@
...
+    {
+        $this->client->addRequestData('rating', $rating);
+    }
+
+    /**
+     * @When I accept the :productReview product review
+     */
+    public function iAcceptTheProductReview(ReviewInterface $productReview): void
+    {
+        $this->client->applyTransition('product_reviews',(string) $productReview->getId(), ProductReviewTransitions::TRANSITION_ACCEPT);
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Api/Admin/ManagingProductReviewsContext.php]: Missing space before `(string)`

File: src/Sylius/Behat/Context/Api/Admin/ManagingProductReviewsContext.php [hunk 17]
```diff
@@ -0,0 +1,237 @@
...
+    }
+
+    /**
+     * @When I reject the :productReview product review
+     */
+    public function iRejectTheProductReview(ReviewInterface $productReview): void
+    {
+        $this->client->applyTransition('product_reviews', (string) $productReview->getId(), ProductReviewTransitions::TRANSITION_REJECT);
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Api/Admin/ManagingProductReviewsContext.php]: What would you say to merge this step the previous one - `I (accept|reject) the :productReview review`? I know we're then not using consts, but IMO we can deal with this danger, as it's a Behat code and not extended by anyone 💃

File: src/Sylius/Behat/Context/Api/Admin/ManagingProductReviewsContext.php [hunk 18]
```diff
@@ -0,0 +1,237 @@
...
+    /**
+     * @Then /^I should be notified that it has been successfully (accepted|rejected)$/
+     */
+    public function iShouldBeNotifiedThatItHasBeenSuccessfullyUpdated(string $action): void
+    {
+        $this->assertIfReviewHasElementWithValue($this->sharedStorage->get('product_review'), 'status', $action);
+    }
+
+    /**
+     * @When I delete the :productReview product review
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Api/Admin/ManagingProductReviewsContext.php]: Please, reorder steps implementations to have them in `Given` -> `When` -> `Then` order 😄

File: src/Sylius/Behat/Resources/config/suites/api/product/managing_product_reviews.yml [hunk 19]
```diff
@@ -3,19 +3,25 @@
 
 default:
     suites:
-        api_managing_product_variants:
+        api_managing_product_reviews:
```
Comment (Zales0123) [inline src/Sylius/Behat/Resources/config/suites/api/product/managing_product_reviews.yml]: Why did you remove the previous suite? 😱

File: src/Sylius/Bundle/ApiBundle/Applicator/ProductReviewStateMachineTransitionApplicator.php [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+        return $data;
+    }
+
+    public function reject(ReviewInterface $data): ReviewInterface
+    {
+        $stateMachine = $this->stateMachineFactory->get($data, ProductReviewTransitions::GRAPH);
+
+        $stateMachine->apply(ProductReviewTransitions::TRANSITION_REJECT);
+
+        return $data;
```
Comment (Zales0123) [inline src/Sylius/Bundle/ApiBundle/Applicator/ProductReviewStateMachineTransitionApplicator.php]: Similar implementation to the previous one, could have some private method extracted 🚀

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Customer.xml [hunk 21]
```diff
@@ -0,0 +1,25 @@
...
+        class="Sylius\Component\Core\Model\Customer"
+        shortName="Customer"
+    >
+        <attribute name="normalization_context">
+            <attribute name="groups">
+                <attribute>customer:read</attribute>
+            </attribute>
+        </attribute>
+        <itemOperations>
+            <itemOperation name="get">
```
Comment (Zales0123) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Customer.xml]: As I understand, it's required to have `Customer`'s (and `Product`s) IRI embedded within a `ProductReview`?

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/ProductReview.xml (hunk 3 above)
Comment (Zales0123) [reply src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/ProductReview.xml]: How does it look right now?

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Customer.xml (hunk 21 above)
Comment (AdamKasp) [reply src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Customer.xml]: yes, without this basic configuration APi platform serialize all related entity :( I hope this is answer for your question :D

Comment (AdamKasp) [reply src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Customer.xml]: yes, without this basic configuration APi platform serialize all related entity :( I hope this is answer for your question :D

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/ProductReview.xml (hunk 3 above)
Comment (AdamKasp) [reply src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/ProductReview.xml]: Screenshot 2020-03-05 at 14 58 12

but as I sad wy should have an empty body

File: src/Sylius/Behat/Context/Api/Admin/ManagingProductReviewsContext.php (hunk 17 above)
Comment (AdamKasp) [reply src/Sylius/Behat/Context/Api/Admin/ManagingProductReviewsContext.php]: I will fix it in next PR

File: src/Sylius/Bundle/ApiBundle/Applicator/ProductReviewStateMachineTransitionApplicator.php [hunk 22]
```diff
@@ -0,0 +1,49 @@
...
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\Applicator;
+
+use SM\Factory\FactoryInterface as StateMachineFactoryInterface;
+use Sylius\Component\Core\ProductReviewTransitions;
+use Sylius\Component\Review\Model\ReviewInterface;
+
+final class ProductReviewStateMachineTransitionApplicator
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Applicator/ProductReviewStateMachineTransitionApplicator.php]: This class could be speced

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/ProductReview.xml (hunk 3 above)
Comment (lchrusciel) [reply src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/ProductReview.xml]: It is not possible to update any of these fields according to https://github.com/Sylius/Sylius/pull/11173/files#diff-8705efddce0ceb9c7db5b7c1aa26bdc9R30-R35 (at least I think so). 

Something, that should be reiterated later, but we can accept the current solution right now.

File: src/Sylius/Behat/Context/Api/Admin/ManagingProductReviewsContext.php [hunk 23]
```diff
@@ -0,0 +1,231 @@
...
+     */
+    public function iAcceptTheProductReview( ReviewInterface $productReview): void
+    {
+        $this->client->applyTransition('product_reviews', (string) $productReview->getId(), 'accept');
+    }
+
+    /**
+     * @When I reject the :productReview product review
+     */
+    public function iRejectTheProductReview( ReviewInterface $productReview): void
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Api/Admin/ManagingProductReviewsContext.php]: Extra space before `ReviewInterface`, here and above

File: src/Sylius/Behat/Client/ApiClientInterface.php [hunk 24]
```diff
@@ -25,7 +25,8 @@ public function buildCreateRequest(string $resource): void;
 
     public function buildUpdateRequest(string $resource, string $id): void;
 
-    public function addRequestData(string $key, string $value): void;
+    /** @param string|int */
```
Comment (GSadee) [inline src/Sylius/Behat/Client/ApiClientInterface.php]:
```suggestion
    /** @param string|int $value */
```

File: src/Sylius/Behat/Client/ApiPlatformClient.php [hunk 25]
```diff
@@ -97,6 +98,16 @@ public function delete(string $resource, string $id): void
         $this->request('DELETE', sprintf('/new-api/%s/%s', $resource, $id), []);
     }
 
+    public function applyTransition(string $resource, string $id, string $transition): void
+    {
+        $this->request(
+            'PATCH',
+            sprintf('/new-api/%s/%s/apply_transition/%s', $resource, $id, $transition),
```
Comment (GSadee) [inline src/Sylius/Behat/Client/ApiPlatformClient.php]:
```suggestion
            sprintf('/new-api/%s/%s/apply-transition/%s', $resource, $id, $transition),
```
But $resource could have `_`, so do we have any convention?

File: src/Sylius/Behat/Context/Api/Admin/ManagingProductReviewsContext.php [hunk 26]
```diff
@@ -0,0 +1,231 @@
...
+use Sylius\Component\Core\ProductReviewTransitions;
+use Sylius\Component\Review\Model\ReviewInterface;
+use Webmozart\Assert\Assert;
+
+final class ManagingProductReviewsContext implements Context
+{
+    /** @var ApiClientInterface */
+    private $client;
+
+    /** @var SharedStorage */
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Admin/ManagingProductReviewsContext.php]:
```suggestion
    /** @var SharedStorageInterface */
```

File: src/Sylius/Behat/Context/Api/Admin/ManagingProductReviewsContext.php [hunk 27]
```diff
@@ -0,0 +1,231 @@
...
+     */
+    public function iWantToBrowseProductReviews(): void
+    {
+        $this->client->index('product_reviews');
+    }
+
+    /**
+     * @When I want to modify the :productReview product review
+     */
+    public function iWantToModifyTheProductReview(ReviewInterface $productReview): void
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Admin/ManagingProductReviewsContext.php]: Could be
```suggestion
    public function iWantToModifyTheProductReview(ProductReviewInterface $productReview): void
```

File: src/Sylius/Behat/Context/Api/Admin/ManagingProductReviewsContext.php [hunk 28]
```diff
@@ -0,0 +1,231 @@
...
+    {
+        $this->assertIfReviewHasElementWithValue($productReview, 'comment', $comment);
+    }
+
+    /**
+     * @Then /^average rating of (product "[^"]+") should be (\d+)$/
+     */
+    public function thisProductAverageRatingShouldBe(ProductInterface $product, int $averageRating): void
+    {
+        // TODO, this step should be in product context which will be covered in separate RP
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Admin/ManagingProductReviewsContext.php]: Probably this step should not be added

File: src/Sylius/Bundle/ApiBundle/Applicator/ProductReviewStateMachineTransitionApplicator.php [hunk 29]
```diff
@@ -0,0 +1,49 @@
...
+{
+    /** @var StateMachineFactoryInterface $stateMachineFactory */
+    private $stateMachineFactory;
+
+    public function __construct(StateMachineFactoryInterface $stateMachineFactory)
+    {
+        $this->stateMachineFactory = $stateMachineFactory;
+    }
+
+    public function accept(ReviewInterface $data): ReviewInterface
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Applicator/ProductReviewStateMachineTransitionApplicator.php]:
```suggestion
    public function accept(ReviewInterface $review): ReviewInterface
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/services.xml [hunk 30]
```diff
@@ -12,5 +12,10 @@
 -->
 
 <container xmlns="http://symfony.com/schema/dic/services" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
-    <services />
+
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/services.xml]: Unnecessary blank line

**Comment by GSadee** [comment]
Thanks, Adam! :tada:

File: src/Sylius/Behat/Context/Api/Admin/ManagingProductReviewsContext.php (hunk 27 above)
Comment (AdamKasp) [reply src/Sylius/Behat/Context/Api/Admin/ManagingProductReviewsContext.php]: I'm not sure because we don't have this interface :D
`ProductReview`` extends `Review` which implements `ReviewInterface` :)

**Comment by AdamKasp** [comment]
all suggestions a fixed in https://github.com/Sylius/Sylius/pull/11197

**Comment by lchrusciel** [comment]
Part of #11250
